patches/server/0966-Call-BlockGrowEvent-for-missing-blocks.patch

```diff
@@ -0,0 +1,176 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
+Date: Fri, 9 Jun 2023 13:04:42 +0200
+Subject: [PATCH] Call BlockGrowEvent for missing blocks
+
+Fixes some cancellability issues for sugar cane, cactus and melon/pumkin stems
+Include the source block when available
+
+diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
+index 55026e1731e41b4e3e4c6a8fef5d96a32051a556..98d1104fa42512264d72ddc3c9801a44013091ab 100644
+--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
++++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
+@@ -1247,8 +1247,8 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
+                         }
+ 
+                         if (iblockdata1 != null && CraftEventFactory.callEntityChangeBlockEvent(Bee.this, blockposition, iblockdata1)) { // CraftBukkit
++                            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(Bee.this.level(), blockposition, iblockdata1, 3)) continue; // Paper
+                             Bee.this.level().levelEvent(2005, blockposition, 0);
+-                            Bee.this.level().setBlockAndUpdate(blockposition, iblockdata1);
+                             Bee.this.incrementNumCropsGrownSincePollination();
+                         }
+                     }
+diff --git a/src/main/java/net/minecraft/world/level/block/CactusBlock.java b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
+index 0003fb51ae3a6575575e10b4c86719f3061e2577..c02f7b0dbf8aaafa76da99a76f1ff87e1a1f0274 100644
+--- a/src/main/java/net/minecraft/world/level/block/CactusBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
+@@ -59,13 +59,13 @@ public class CactusBlock extends Block {
+ 
+                 int modifier = world.spigotConfig.cactusModifier; // Spigot - SPIGOT-7159: Better modifier resolution
+                 if (j >= 15 || (modifier != 100 && random.nextFloat() < (modifier / (100.0f * 16)))) { // Spigot - SPIGOT-7159: Better modifier resolution
+-                    CraftEventFactory.handleBlockGrowEvent(world, blockposition1, this.defaultBlockState()); // CraftBukkit
++                    if (!CraftEventFactory.handleBlockGrowEvent(world, blockposition1, pos, this.defaultBlockState(), 3)) return; // CraftBukkit // Paper - properly cancel
+                     BlockState iblockdata1 = (BlockState) state.setValue(CactusBlock.AGE, 0);
```

Comment:
I think that the `BlockState` argument to `handleBlockGrowEvent` should be the `iblockdata1` that is below instead of the returned getDefaultBlockState. They should always be the same now, but what if cactus somehow get another blockstate property or whatever.

Reply:
the blockstate below is the sourceblock but the grow event called when a new block grow always have the new block data instead which is the defaultBlockState see here: https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/nms-patches/net/minecraft/world/level/block/BlockCactus.patch#17 the event call replace the setblock